### DISCUSSION TITLE
Hedera Token Service Precompiled Contract Token Management Functions

### DISCUSSION THREAD

**Discussion by mgarbs** [hiero-ledger/hiero-improvement-proposals#516] [Ideas]
This discussion is to discuss HIP-514 - Hedera Token Service Precompiled Contract Token Management Functions

## Abstract

The Hedera Smart Contract Service (HSCS), through the initial implementation of [HIP 206](https://hips.hedera.com/hip/hip-206), allows contracts to create, transfer, mint, burn, associate, and dissociate tokens programmatically. However, there remains limited capability to interact with and manage tokens in more intricate scenarios.

## Motivation

Allow smart contract developers on Hedera similar capabilities in token management as developers using HTS.

Will add a link to the HIP when it becomes available.

@Nana-EC

**Comment by stoqnkpL** [comment]
Suggestion: add one more function to HIP-514:
`exists (address _token) returns (uint64, bool)`
which validates whether an address is a valid token address and the token is not deleted.
It will improve DevEx for people using the HTS precompiled contract

**Reply by shemnon** [reply]
The name needs more clarity, as solidity function names have basically no namespacing.

perhaps `isToken(address _token) returns (uint64 statusCode, bool isToken)`?

Or a more general `getTokenType(address _token) returns (uint64 statusCode, int32 tokenType)` where tokenType is the `TokenType` message from the protobufs, adding in -1 for "not a token".

**Reply by stoqnkpL** [reply]
I think it makes sense to have both:
- `isToken(address _token) returns (uint64 statusCode, bool isToken)`
- `getTokenType(address _token) returns (uint64 statusCode, int32 tokenType)`
They can be used in different contexts.

**Reply by shemnon** [reply]
Both is acceptable.
